md2man/roff.go

```diff
@@ -1,252 +1,281 @@
-	case level == 1:
-		out.WriteString("\n\n.SH ")
-	case level == 2:
-		out.WriteString("\n.SH ")
-	default:
-		out.WriteString("\n.SS ")
-	}
-
-	if !text() {
-		out.Truncate(marker)
-		return
+func NewRoffRenderer() *RoffRenderer {
+	var extensions blackfriday.Extensions
+
+	extensions |= blackfriday.NoIntraEmphasis
+	extensions |= blackfriday.Tables
+	extensions |= blackfriday.FencedCode
+	extensions |= blackfriday.SpaceHeadings
+	extensions |= blackfriday.Footnotes
+	extensions |= blackfriday.Titleblock
+	extensions |= blackfriday.DefinitionLists
+	return &RoffRenderer{
+		extensions: extensions,
 	}
 }
 
-func (r *roffRenderer) HRule(out *bytes.Buffer) {
-	out.WriteString("\n.ti 0\n\\l'\\n(.lu'\n")
+// GetExtensions returns the list of extensions used by this renderer implementation
+func (r *RoffRenderer) GetExtensions() blackfriday.Extensions {
+	return r.extensions
 }
 
-func (r *roffRenderer) List(out *bytes.Buffer, text func() bool, flags int) {
-	marker := out.Len()
-	r.ListCounters = append(r.ListCounters, 1)
-	out.WriteString("\n.RS\n")
-	if !text() {
-		out.Truncate(marker)
-		return
-	}
-	r.ListCounters = r.ListCounters[:len(r.ListCounters)-1]
-	out.WriteString("\n.RE\n")
-}
-
-func (r *roffRenderer) ListItem(out *bytes.Buffer, text []byte, flags int) {
-	if flags&blackfriday.LIST_TYPE_ORDERED != 0 {
-		out.WriteString(fmt.Sprintf(".IP \"%3d.\" 5\n", r.ListCounters[len(r.ListCounters)-1]))
-		r.ListCounters[len(r.ListCounters)-1]++
-	} else {
-		out.WriteString(".IP \\(bu 2\n")
+// RenderHeader handles outputting the header at document start
+func (r *RoffRenderer) RenderHeader(w io.Writer, ast *blackfriday.Node) {
+	// disable hyphenation
+	if _, err := io.WriteString(w, ".nh\n"); err != nil {
+		fmt.Fprintf(os.Stderr, "ERROR: Unable to output header string to writer: %v\n", err)
 	}
-	out.Write(text)
-	out.WriteString("\n")
 }
 
-func (r *roffRenderer) Paragraph(out *bytes.Buffer, text func() bool) {
-	marker := out.Len()
-	out.WriteString("\n.PP\n")
-	if !text() {
-		out.Truncate(marker)
-		return
-	}
-	if marker != 0 {
-		out.WriteString("\n")
-	}
+// RenderFooter handles outputting the footer at the document end; the roff
+// renderer has no footer information
+func (r *RoffRenderer) RenderFooter(w io.Writer, ast *blackfriday.Node) {
 }
 
-func (r *roffRenderer) Table(out *bytes.Buffer, header []byte, body []byte, columnData []int) {
-	out.WriteString("\n.TS\nallbox;\n")
+// RenderNode is called for each node in a markdown document; based on the node
+// type the equivalent roff output is sent to the writer
+func (r *RoffRenderer) RenderNode(w io.Writer, node *blackfriday.Node, entering bool) blackfriday.WalkStatus {
 
-	maxDelims := 0
-	lines := strings.Split(strings.TrimRight(string(header), "\n")+"\n"+strings.TrimRight(string(body), "\n"), "\n")
-	for _, w := range lines {
-		curDelims := strings.Count(w, "\t")
-		if curDelims > maxDelims {
-			maxDelims = curDelims
+	switch node.Type {
+	case blackfriday.Text:
```

Comment:
Can we break out the switch cases that are more complicated than a simple `out()` call?